edgelet/edgelet-kube/src/runtime.rs

```diff
@@ -432,9 +477,60 @@ where
     fn remove_all(&self) -> Self::RemoveAllFuture {
         Box::new(future::ok(()))
     }
+}
+
+impl<T, S> Authenticator for KubeModuleRuntime<T, S>
+where
+    T: TokenSource + Clone + 'static,
+    S: Service + Send + 'static,
+    S::ReqBody: From<Vec<u8>>,
+    S::ResBody: Stream,
+    Body: From<S::ResBody>,
+    <S::ResBody as Stream>::Item: AsRef<[u8]>,
+    <S::ResBody as Stream>::Error: Into<KubeClientError>,
+    S::Error: Into<KubeClientError>,
+    S::Future: Send,
+{
+    type Error = Error;
+    type Request = Request<Body>;
+    type AuthenticateFuture = Box<dyn Future<Item = AuthId, Error = Self::Error> + Send>;
+
+    fn authenticate(&self, req: &Self::Request) -> Self::AuthenticateFuture {
+        let token = req
+            .headers()
+            .get(header::AUTHORIZATION)
+            .and_then(|token| token.to_str().ok())
+            .filter(|token| token.len() > 6 && &token[..7].to_uppercase() == "BEARER ")
+            .map(|token: &str| &token[7..]);
+
+        let fut = match token {
+            Some(token) => Either::A(
+                self.client
+                    .lock()
+                    .expect("Unexpected lock error")
+                    .borrow_mut()
+                    .token_review(token)
+                    .map_err(|err| {
+                        log_failure(Level::Warn, &err);
+                        Error::from(err)
+                    })
+                    .and_then(|token_review| {
```

Comment:
It may be possible to write this as:
```rust
.map(|token_review| {
    token_review
        .status
        .as_ref()
        .filter(|status| status.authenticated.filter(|x| *x).is_some())
        .and_then(|status| {
            status.user.as_ref().and_then(|user| user.username.clone())
        })
        .map_or(AuthId::None, AuthId::Value)
}),
```

Reply:
Sure